compiler/ilgen/VirtualMachineState.hpp

```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2016, 2016
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+
+#ifndef OMR_VIRTUALMACHINESTATE_INCL
+#define OMR_VIRTUALMACHINESTATE_INCL
+
+
+namespace TR { class IlBuilder; }
+class TR_Memory;
+
+template <class T> class List;
+template <class T> class ListAppender;
+
+namespace OMR
+{
+
+/**
+ * @brief Interface for expressing virtual machine state variables to JitBuilder
+ *
+ * VirtualMachineState is an interface that language compilers will implement
+ * to define all the aspects of the virtual machine state that the JIT compiler will
+ * simulate (operand stacks, virtual registers like "pc", "sp", etc.). The interface
+ * has four functions: 1) Commit() to cause the simulated state to be written to the
+ * actual virtual machine state, 2) Reload() to set the simulated machine state based
+ * on the current virtual machine state, 3) MakeCopy() to make an identical copy of
+ * the curent virtual machine state object, and 4) MergeInto() to perform the actions
+ * needed for the current state to be merged with a second virtual machine state.
+ *
+ * ##Usage
+ *
+ * Commit() is typically called when transitioning from compiled code to the interpreter.
+ *
+ * Reload() is typically called on the transition back from the interpreter to compiled
+ * code.
+ *
+ * MakeCopy() is typically called when control flow edges are created: the current vm state
+ * must also become the initial vm state at the target of the control flow edge and be
+ * able to evolve independently from the current vm state.
```

Comment:
Yes. Otherwise, any change to the state in the "source" builder object would also change the state in the "target" builder object. On top of that, every builder object needs to maintain the "initial" vm state (i.e. the state when control first gets there) as well as the "current" vm state (i.e. the state as it evolves due to the operations performed by that builder). You need the initial state so that any MergeInto() operations to that builder object synchronize to the proper vm state...